compiler/rustc_lexer/src/lib.rs

```diff
@@ -758,6 +806,34 @@ impl Cursor<'_> {
         false
     }
 
+    /// Eats the double-quoted string and returns `n_hashes` and an error if encountered.
+    fn guarded_double_quoted_string(&mut self, n_start_hashes: u32) -> Result<u32, RawStrError> {
+        debug_assert!(self.prev() == '"');
+
+        // Lex the string itself as a normal string literal
+        // so we can recover that for older editions later.
+        if !self.double_quoted_string() {
+            return Err(RawStrError::NoTerminator {
+                expected: n_start_hashes,
+                found: 0,
+                possible_terminator_offset: None,
+            });
+        }
+
+        // Consume closing '#' symbols.
+        // Note that this will not consume extra trailing `#` characters:
+        // `###"abcde"####` is lexed as a `GuardedStr { n_hashes: 3 }`
+        // followed by a `#` token.
+        let mut n_end_hashes = 0;
+        while self.first() == '#' && n_end_hashes < n_start_hashes {
+            n_end_hashes += 1;
+            self.bump();
+        }
+
+        // Handle `n_end_hashes < n_start_hashes` later.
```

Comment:
I can't see any checking that the number of start and end hashes match. Am I overlooking it?